repos/system_upgrade/el8toel9/actors/checkifcfg/libraries/ifcfg.py

```diff
@@ -0,0 +1,145 @@
+import os
+
+from leapp import reporting
+
+
+def process():
+    TRUE_VALUES = ['yes', 'true', '1']
+    GOOD_TYPES = [
+        'ethernet', 'ctc', 'wireless', 'infiniband', 'bond',
+        'team', 'vlan', 'bridge'
+    ]
+
+    rule_files = []
+    bad_type_files = []
+    not_controlled_files = []
+
+    try:
+        files = os.scandir('/etc/sysconfig/network-scripts')
+    except OSError:
+        # The legacy network-scripts package is probably already
+        # gone without a trace. Good!
+        return
+
+    for f in files:
+        bad_type = False
+        got_type = None
+        nm_controlled = True
+
+        if not f.is_file():
+            continue
+
+        if f.name.startswith('rule-') or f.name.startswith('rule6-'):
+            rule_files.append(f.path)
+            continue
+
+        if not f.name.startswith('ifcfg-'):
+            continue
+
+        if f.name == 'ifcfg-lo':
+            continue
+
+        for line in open(f.path):
+            try:
+                (key, value) = line.split('#')[0].strip().split('=')
+            except ValueError:
+                # We're not interested in lines that are not
+                # simple assignments. Play it safe.
+                continue
```

Comment:
If there are complex structures that cannot be parsed, it seems it would be best to inhibit the upgrade, wouldn't it?

Reply:
It's not necessarily complex structures. It's anything multi-line, comments, etc.
We quite intentionally don't implement a full-fledged parser here, because 1.) it's very difficult to get right 2.) we really only need to read TYPE and NM_CONTROLLED. We're not in the position to judge whether NM would be able to load the file here. Also, NM's parser also just ignores things it can't grok.

Reply:
I agree with your observations. The consequences are different for me, though. If there are indications that NM might ignore something, then the upgrade should be blocked IMHO.